File: backend/app/api/routes/analyses.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.schemas.analysis import (
    AnalysisCreateRequest,
    AnalysisDetailResponse,
    AnalysisFeedbackRequest,
    AnalysisListResponse,
    AnalysisRecalculateRequest,
    RecommendationHistoryResponse,
)
from app.services.analysis_service import AnalysisService

router = APIRouter(prefix="/analyses", tags=["analyses"])


@router.post("", response_model=AnalysisDetailResponse)
async def create_analysis(
    payload: AnalysisCreateRequest, db: AsyncSession = Depends(get_db)
):
    return await AnalysisService(db).create_analysis(payload)


@router.get("", response_model=AnalysisListResponse)
async def list_analyses(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=10, ge=1, le=50),
    city: str | None = None,
    property_type: str | None = None,
    search: str | None = None,
    db: AsyncSession = Depends(get_db),
):
    return await AnalysisService(db).list_analyses(page, page_size, city, property_type, search)


@router.get("/{analysis_id}", response_model=AnalysisDetailResponse)
async def get_analysis(analysis_id: UUID, db: AsyncSession = Depends(get_db)):
    return await AnalysisService(db).get_analysis(analysis_id)


@router.post("/{analysis_id}/feedback", response_model=AnalysisDetailResponse)
async def add_feedback(
    analysis_id: UUID,
    payload: AnalysisFeedbackRequest,
    db: AsyncSession = Depends(get_db),
):
    return await AnalysisService(db).add_feedback(analysis_id, payload)


@router.post("/{analysis_id}/recalculate", response_model=AnalysisDetailResponse)
async def recalculate_analysis(
    analysis_id: UUID,
    payload: AnalysisRecalculateRequest,
    db: AsyncSession = Depends(get_db),
):
    return await AnalysisService(db).recalculate_analysis(analysis_id, payload)


@router.get("/{analysis_id}/history", response_model=list[RecommendationHistoryResponse])
async def get_analysis_history(
    analysis_id: UUID, db: AsyncSession = Depends(get_db)
):
    return await AnalysisService(db).get_history(analysis_id)

File: frontend/services/analyses.ts
import type { AnalysisDetail, AnalysisListResponse } from "@/types/analysis";
import type { RecommendationPayload, UpdateRecommendationPayload } from "@/types/pricing";

const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000/api/v1";

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      ...(init?.headers ?? {}),
    },
    cache: "no-store",
  });
  if (!response.ok) {
    const message = await response.text();
    throw new Error(message || `Request failed: ${response.status}`);
  }
  return response.json() as Promise<T>;
}

export async function createAnalysis(
  payload: RecommendationPayload,
): Promise<AnalysisDetail> {
  return request("/analyses", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function recalculateAnalysis(
  analysisId: string,
  payload: UpdateRecommendationPayload,
): Promise<AnalysisDetail> {
  return request(`/analyses/${analysisId}/recalculate`, {
    method: "POST",
    body: JSON.stringify({
      feedback_text: payload.feedbackText,
      comparable_actions: payload.comparableActions?.map((action) => ({
        comparable_id: action.comparableId,
        action: action.action,
      })),
    }),
  });
}

export async function listAnalyses(params?: {
  page?: number;
  pageSize?: number;
  city?: string;
  propertyType?: string;
  search?: string;
}): Promise<AnalysisListResponse> {
  const searchParams = new URLSearchParams();
  if (params?.page) searchParams.set("page", String(params.page));
  if (params?.pageSize) searchParams.set("page_size", String(params.pageSize));
  if (params?.city) searchParams.set("city", params.city);
  if (params?.propertyType) searchParams.set("property_type", params.propertyType);
  if (params?.search) searchParams.set("search", params.search);
  const suffix = searchParams.toString() ? `?${searchParams.toString()}` : "";
  return request(`/analyses${suffix}`);
}

export async function getAnalysis(analysisId: string): Promise<AnalysisDetail> {
  return request(`/analyses/${analysisId}`);
}

export async function getAnalysisHistory(analysisId: string) {
  return request(`/analyses/${analysisId}/history`);
}
